
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../pt-3/">
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.18">
    
    
      
        <title>Practical 4 - BIO773P</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.7e37652d.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#practical-4-computing-in-r" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="BIO773P" class="md-header__button md-logo" aria-label="BIO773P" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            BIO773P
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Practical 4
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../pt-1/" class="md-tabs__link">
          
  
  
  Practicals

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="BIO773P" class="md-nav__button md-logo" aria-label="BIO773P" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    BIO773P
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Programming in R
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Practicals
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Practicals
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pt-1/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Practical 1
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pt-2/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Practical 2
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pt-3/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Practical 3
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Practical 4
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Practical 4
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#recap-questions" class="md-nav__link">
    <span class="md-ellipsis">
      Recap questions
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Recap questions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#intro-q1-what-would-be-the-outcome-of-the-code-answer-repx-c42-24-times-42-and-what-would-then-be-the-outcome-of-the-code-meananswer" class="md-nav__link">
    <span class="md-ellipsis">
      Intro Q1. What would be the outcome of the code answer &lt;- rep(x = c(42, 24), times = 42)? And what would then be the outcome of the code mean(answer)?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#intro-q2-use-the-preloaded-r-data-set-indometh-check-indometh-to-understand-more-about-the-dataset-subset-this-data-to-return-only-those-rows-for-which-the-concentration-is-strictly-between-1-and-2-what-is-the-average-mean-concentration-for-this-subset" class="md-nav__link">
    <span class="md-ellipsis">
      Intro Q2. Use the preloaded R data set 'Indometh' (check ?Indometh to understand more about the dataset). Subset this data to return only those rows for which the concentration is strictly between 1 and 2. What is the average (mean) concentration for this subset?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#intro-q3-using-regular-expressions-how-would-you-extract-all-the-words-except-antelope-in-the-vector-ccameleopard-eop4a-kiloparsec-antelope-in-r-there-are-a-few-ways-of-getting-a-result-find-three-ways-to-answer-this-question" class="md-nav__link">
    <span class="md-ellipsis">
      Intro Q3. Using regular expressions, how would you extract all the words except antelope in the vector c("cameleopard", "eop4a", "kiloparsec", "antelope")? In R, there are a few ways of getting a result. Find three ways to answer this question.
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#intro-q4-write-a-loop-that-iterates-over-the-numbers-16-to-49-and-prints-out-the-square-root-of-the-number-each-time-through-you-may-have-to-search-around-for-the-square-root-function-yes-for-this-exercise-we-absolutely-want-you-to-write-a-loop-to-do-this-even-though-there-are-other-ways-of-getting-r-to-provide-the-same-result" class="md-nav__link">
    <span class="md-ellipsis">
      Intro Q4. Write a loop that iterates over the numbers 16 to 49 and prints out the square root of the number each time through (you may have to search around for the square root function). Yes, for this exercise we absolutely want you to write a loop to do this, even though there are other ways of getting R to provide the same result.
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#intro-q5-make-the-loop-from-intro-q4-store-the-results-to-a-separate-vector-called-my_square_roots-instead-of-just-printing-the-results-whats-the-value-of-the-3rd-iteration-what-is-the-sum-of-the-square-roots-of-the-numbers-16-to-49" class="md-nav__link">
    <span class="md-ellipsis">
      Intro Q5. Make the loop from Intro Q4 store the results to a separate vector called my_square_roots instead of just printing the results. What's the value of the 3rd iteration? What is the sum of the square roots of the numbers 16 to 49?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#intro-q6-write-a-loop-that-iterates-over-all-even-numbers-between-30-and-90-at-each-iteration-run-your-function-to-convert-each-of-these-even-numbers-from-kilometres-into-miles-you-created-this-function-in-q2-from-practical2-the-loop-should-store-the-results-in-a-separate-vector" class="md-nav__link">
    <span class="md-ellipsis">
      Intro Q6. Write a loop that iterates over all even numbers between 30 and 90. At each iteration, run your function to convert each of these even numbers from kilometres into miles (you created this function in Q2 from practical2. The loop should store the results in a separate vector.
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#intro-q7-numbers-in-fibonaccis-sequence-are-characterised-by-the-fact-that-every-number-after-the-first-two-is-the-sum-of-the-two-preceding-ones-eg-1-1-2-3-5-this-seemingly-simple-sequence-can-be-used-to-explain-different-biological-patterns-ranging-from-the-arrangement-of-leaves-on-a-stem-or-the-flowering-of-an-artichoke-to-the-patterns-of-inheritance-of-x-chromosomes-write-a-for-loop-that-calculates-the-first-20-numbers-in-a-fibonacci-sequence-starting-from-1" class="md-nav__link">
    <span class="md-ellipsis">
      Intro Q7. Numbers in Fibonacci's sequence are characterised by the fact that every number after the first two is the sum of the two preceding ones (e.g., 1, 1, 2, 3, 5,...). This seemingly simple sequence can be used to explain different biological patterns ranging from the arrangement of leaves on a stem or the flowering of an artichoke to the patterns of inheritance of X chromosomes. Write a for loop that calculates the first 20 numbers in a Fibonacci sequence starting from 1.
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#intro-q8-create-a-function-that-can-run-this-loop-to-generate-a-fibonacci-sequence-of-any-size-starting-from-any-given-number-ie-it-does-not-have-to-start-with-1-so-for-example-a-sequence-of-length-x-from-number-y-where-x-and-y-are-inputs-to-the-function" class="md-nav__link">
    <span class="md-ellipsis">
      Intro Q8. Create a function that can run this loop to generate a Fibonacci sequence of any size, starting from any given number (i.e. it does not have to start with 1). So for example, a sequence of length x from number y, where x and y are inputs to the function.
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#palindromic-sequences" class="md-nav__link">
    <span class="md-ellipsis">
      Palindromic Sequences
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Palindromic Sequences">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#q1-write-a-function-that-assesses-whether-a-given-word-or-phrase-is-a-palindrome" class="md-nav__link">
    <span class="md-ellipsis">
      Q1. Write a function that assesses whether a given word or phrase is a palindrome.
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#q2-palindromes-in-molecular-biology-work-a-little-differently-in-that-the-sequence-is-the-same-if-read-from-5-to-3-on-one-strand-or-from-5-to-3-on-the-complementary-strand-for-example-given-the-sequence-acctaggt-its-complement-would-be" class="md-nav__link">
    <span class="md-ellipsis">
      Q2. Palindromes in molecular biology work a little differently, in that the sequence is the same if read from 5’ to 3’ on one strand or from 5’ to 3’ on the complementary strand. For example, given the sequence ACCTAGGT, its complement would be:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#both-acctaggt-and-its-complementary-sequence-tggatcca-are-identical-if-read-from-5-to-3-that-is-acctaggt-is-a-palindromic-sequence-write-a-function-that-assesses-whether-a-dna-sequence-is-a-palindrome-you-will-need-to-convert-it-to-its-reverse-complement-and-then-compare-this-to-the-original-sequence-again-there-are-commands-which-may-help-you-in-the-working-with-dna-data-section-from-the-last-session" class="md-nav__link">
    <span class="md-ellipsis">
      Both ACCTAGGT and its complementary sequence TGGATCCA are identical if read from 5’ to 3’, that is, ACCTAGGT is a palindromic sequence. Write a function that assesses whether a DNA sequence is a palindrome. You will need to convert it to its reverse complement and then compare this to the original sequence. Again, there are commands which may help you in the ‘Working with DNA data’ section from the last session.
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#open-reading-frames" class="md-nav__link">
    <span class="md-ellipsis">
      Open Reading Frames
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Open Reading Frames">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#q3-write-a-function-which-uses-regular-expressions-to-detect-if-a-given-sequence-contains-an-open-reading-frame-test-it-on-the-following-sequences" class="md-nav__link">
    <span class="md-ellipsis">
      Q3. Write a function which uses regular expressions to detect if a given sequence contains an open reading frame. Test it on the following sequences:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#hacker-q4a-instead-of-having-the-function-return-whether-your-sequence-contains-an-open-reading-frame-have-the-function-return-a-string-containing-the-open-reading-frame-if-there-is-one-and-null-if-there-is-not-you-will-need-to-look-up-if-statements" class="md-nav__link">
    <span class="md-ellipsis">
      Hacker Q4a. Instead of having the function return whether your sequence contains an open reading frame, have the function return a string containing the open reading frame if there is one and NULL if there is not (you will need to look up "if statements").
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#hacker-q4b-this-time-check-both-the-forward-and-reverse-strand-for-open-reading-frames-return-forward_only-if-there-is-an-open-reading-frame-only-on-the-forward-strand-reverse_only-if-only-on-the-reverse-both-if-in-both-or-none-if-in-neither" class="md-nav__link">
    <span class="md-ellipsis">
      Hacker Q4b. This time, check both the forward and reverse strand for open reading frames. Return "forward_only" if there is an open reading frame only on the forward strand, "reverse_only" if only on the reverse, "both" if in both or "none" if in neither.
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#species-names" class="md-nav__link">
    <span class="md-ellipsis">
      Species names
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Species names">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#q5-write-a-function-that-adds-a-new-column-in-the-data-frame-that-contains-the-correct-latin-species-name-for-each-record-in-the-butterfly_sample" class="md-nav__link">
    <span class="md-ellipsis">
      Q5 Write a function that adds a new column in the data frame that contains the correct Latin species name for each record in the butterfly_sample.
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#q6-from-the-data-above-which-location-has-the-greatest-number-of-different-species" class="md-nav__link">
    <span class="md-ellipsis">
      Q6. From the data above, which location has the greatest number of different species?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#q7-which-genus-has-been-caught-the-greatest-number-of-times" class="md-nav__link">
    <span class="md-ellipsis">
      Q7. Which genus has been caught the greatest number of times?
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="practical-4-computing-in-r"><strong>Practical 4 - Computing in R</strong></h1>
<hr />
<h2 id="introduction">Introduction</h2>
<p>This session will be a recap of what we saw in <a href="../practical1">practical 1</a> and <a href="../practical2">practical 2</a>. You can also use this session to revisit the parts of the last practicals that are still not entirely clear.
<strong>REMEMBER</strong> The important thing is that you understand what you are doing. It is better to understand well a few exercises than finishing all exercises without entirely knowing what is going on.</p>
<h2 id="recap-questions">Recap questions</h2>
<p>To get back into gear in terms of thinking about programming, we’ll start with a quick recap exercise of the recent material:</p>
<h4 id="intro-q1-what-would-be-the-outcome-of-the-code-answer-repx-c42-24-times-42-and-what-would-then-be-the-outcome-of-the-code-meananswer">Intro Q1. What would be the outcome of the code <code>answer &lt;- rep(x = c(42, 24), times = 42)</code>? And what would then be the outcome of the code <code>mean(answer)</code>?</h4>
<h4 id="intro-q2-use-the-preloaded-r-data-set-indometh-check-indometh-to-understand-more-about-the-dataset-subset-this-data-to-return-only-those-rows-for-which-the-concentration-is-strictly-between-1-and-2-what-is-the-average-mean-concentration-for-this-subset">Intro Q2. Use the preloaded R data set 'Indometh' (check <code>?Indometh</code> to understand more about the dataset). Subset this data to return only those rows for which the concentration is strictly between 1 and 2. What is the average (mean) concentration for this subset?</h4>
<h4 id="intro-q3-using-regular-expressions-how-would-you-extract-all-the-words-except-antelope-in-the-vector-ccameleopard-eop4a-kiloparsec-antelope-in-r-there-are-a-few-ways-of-getting-a-result-find-three-ways-to-answer-this-question">Intro Q3. Using regular expressions, how would you extract all the words except <code>antelope</code> in the vector <code>c("cameleopard", "eop4a", "kiloparsec", "antelope")</code>? In R, there are a few ways of getting a result. Find three ways to answer this question.</h4>
<h4 id="intro-q4-write-a-loop-that-iterates-over-the-numbers-16-to-49-and-prints-out-the-square-root-of-the-number-each-time-through-you-may-have-to-search-around-for-the-square-root-function-yes-for-this-exercise-we-absolutely-want-you-to-write-a-loop-to-do-this-even-though-there-are-other-ways-of-getting-r-to-provide-the-same-result">Intro Q4. Write a loop that iterates over the numbers 16 to 49 and prints out the square root of the number each time through (you may have to search around for the square root function). Yes, for this exercise we <em>absolutely</em> want you to write a loop to do this, even though there are other ways of getting R to provide the same result.</h4>
<h4 id="intro-q5-make-the-loop-from-intro-q4-store-the-results-to-a-separate-vector-called-my_square_roots-instead-of-just-printing-the-results-whats-the-value-of-the-3rd-iteration-what-is-the-sum-of-the-square-roots-of-the-numbers-16-to-49">Intro Q5. Make the loop from Intro Q4 store the results to a separate vector called <code>my_square_roots</code> instead of just printing the results. What's the value of the 3rd iteration? What is the sum of the square roots of the numbers 16 to 49?</h4>
<h4 id="intro-q6-write-a-loop-that-iterates-over-all-even-numbers-between-30-and-90-at-each-iteration-run-your-function-to-convert-each-of-these-even-numbers-from-kilometres-into-miles-you-created-this-function-in-q2-from-practical2-the-loop-should-store-the-results-in-a-separate-vector">Intro Q6. Write a loop that iterates over all even numbers between 30 and 90. At each iteration, run your function to convert each of these even numbers from kilometres into miles (you created this function in Q2 from <a href="../practical2">practical2</a>. The loop should store the results in a separate vector.</h4>
<h4 id="intro-q7-numbers-in-fibonaccis-sequence-are-characterised-by-the-fact-that-every-number-after-the-first-two-is-the-sum-of-the-two-preceding-ones-eg-1-1-2-3-5-this-seemingly-simple-sequence-can-be-used-to-explain-different-biological-patterns-ranging-from-the-arrangement-of-leaves-on-a-stem-or-the-flowering-of-an-artichoke-to-the-patterns-of-inheritance-of-x-chromosomes-write-a-for-loop-that-calculates-the-first-20-numbers-in-a-fibonacci-sequence-starting-from-1">Intro Q7. Numbers in Fibonacci's sequence are characterised by the fact that every number after the first two is the sum of the two preceding ones (e.g., 1, 1, 2, 3, 5,...). This seemingly simple sequence can be used to explain different biological patterns ranging from the arrangement of leaves on a stem or the flowering of an artichoke to the patterns of inheritance of X chromosomes. Write a <code>for</code> loop that calculates the first 20 numbers in a Fibonacci sequence starting from 1.</h4>
<p><em>Hint - we cannot use positions equal or lower than 0 in a vector. Start with a vector containing the 2 first numbers of the sequence (1 and 1 in this case) and run through a loop starting in position 3 of your sequence to calculate the rest</em></p>
<h4 id="intro-q8-create-a-function-that-can-run-this-loop-to-generate-a-fibonacci-sequence-of-any-size-starting-from-any-given-number-ie-it-does-not-have-to-start-with-1-so-for-example-a-sequence-of-length-x-from-number-y-where-x-and-y-are-inputs-to-the-function">Intro Q8. Create a function that can run this loop to generate a Fibonacci sequence of any size, starting from any given number (i.e. it does not have to start with 1). So for example, a sequence of length <code>x</code> from number <code>y</code>, where <code>x</code> and <code>y</code> are inputs to the function.</h4>
<p><em>Hint - The starting number <code>y</code> will be the first two positions of your resulting vector</em></p>
<h2 id="palindromic-sequences">Palindromic Sequences</h2>
<p>Palindromes are arrangements of words or letters which read the same way whether you read them backwards or forwards, such as the phrase <em>never odd or even</em> (if you remove the spaces). In molecular biology, many restriction enzyme sites are palindromic,</p>
<h4 id="q1-write-a-function-that-assesses-whether-a-given-word-or-phrase-is-a-palindrome">Q1. Write a function that assesses whether a given word or phrase is a palindrome.</h4>
<p>Before starting to code, think about the steps that you would need to go through in order to judge if something is a mirror palindrome.
 * Write these steps as comments in an R script window (or a piece of paper) , then think about how you can tell the computer to execute those steps.
 * Only start writing the code when you have a plan of what you want to do.
 * Write your code as a one-off first - without considering that you'll do a function.
 * Only once you're happy with how things work, sandwich your code into a function format.
 * Use easy test cases where you know the answer in order to check that your function works.</p>
<p>The bottom section of the R help sheets normally have examples of how a command can be used. Sometimes one of these examples will be a way to solve the problem that you are currently working on. The <code>strsplit</code> helpsheet is particularly interesting in relation to this question.</p>
<h4 id="q2-palindromes-in-molecular-biology-work-a-little-differently-in-that-the-sequence-is-the-same-if-read-from-5-to-3-on-one-strand-or-from-5-to-3-on-the-complementary-strand-for-example-given-the-sequence-acctaggt-its-complement-would-be">Q2. Palindromes in molecular biology work a little differently, in that the sequence is the same if read from 5’ to 3’ on one strand or from 5’ to 3’ on the complementary strand. For example, given the sequence ACCTAGGT, its complement would be:</h4>
<div class="highlight"><pre><span></span><code>5’ ACCTAGGT 3’
   ||||||||
3’ TGGATCCA 5’
</code></pre></div>
<h4 id="both-acctaggt-and-its-complementary-sequence-tggatcca-are-identical-if-read-from-5-to-3-that-is-acctaggt-is-a-palindromic-sequence-write-a-function-that-assesses-whether-a-dna-sequence-is-a-palindrome-you-will-need-to-convert-it-to-its-reverse-complement-and-then-compare-this-to-the-original-sequence-again-there-are-commands-which-may-help-you-in-the-working-with-dna-data-section-from-the-last-session">Both ACCTAGGT and its complementary sequence TGGATCCA are identical if read from 5’ to 3’, that is, ACCTAGGT is a palindromic sequence. Write a function that assesses whether a DNA sequence is a palindrome. You will need to convert it to its reverse complement and then compare this to the original sequence. Again, there are commands which may help you in the ‘Working with DNA data’ section from the <a href="../practical2">last session</a>.</h4>
<h2 id="open-reading-frames">Open Reading Frames</h2>
<p>Protein-coding regions in the genome can be predicted by detecting open reading frames. An open reading frame normally begins with the start codon ‘ATG’ and ends at one of three possible stop codons, ‘TGA’, ‘TAA’ and ‘TAG’. The sequence in between these two points is arranged in 3-base codons.</p>
<h4 id="q3-write-a-function-which-uses-regular-expressions-to-detect-if-a-given-sequence-contains-an-open-reading-frame-test-it-on-the-following-sequences">Q3. Write a function which uses regular expressions to detect if a given sequence contains an open reading frame. Test it on the following sequences:</h4>
<p><em>Hint - an open reading frame should always start with the start codon 'ATG' and end with one of the stop codons. Additionally, the length of the entire sequence should be a multiple of 3. If any of these features are not met, the sequence is not a functional open reading frame.</em></p>
<div class="highlight"><pre><span></span><code>ATGGATTTTTAG
ATGGATTTTCTAG
CTAATGGATTTTTGAAT
atgctaaactaa
TCGATTAA
</code></pre></div>
<h4 id="hacker-q4a-instead-of-having-the-function-return-whether-your-sequence-contains-an-open-reading-frame-have-the-function-return-a-string-containing-the-open-reading-frame-if-there-is-one-and-null-if-there-is-not-you-will-need-to-look-up-if-statements">Hacker Q4a. Instead of having the function return whether your sequence contains an open reading frame, have the function return a string containing the open reading frame if there is one and NULL if there is not (you will need to look up "if statements").</h4>
<h4 id="hacker-q4b-this-time-check-both-the-forward-and-reverse-strand-for-open-reading-frames-return-forward_only-if-there-is-an-open-reading-frame-only-on-the-forward-strand-reverse_only-if-only-on-the-reverse-both-if-in-both-or-none-if-in-neither">Hacker Q4b. This time, check both the forward and reverse strand for open reading frames. Return <code>"forward_only"</code> if there is an open reading frame only on the forward strand, <code>"reverse_only"</code> if only on the reverse, <code>"both"</code> if in both or "none" if in neither.</h4>
<h2 id="species-names">Species names</h2>
<p>Run the following line of code to import the <code>butterfly_sample</code> and the <code>butterfly_reference</code> data frames:</p>
<div class="highlight"><pre><span></span><code><span class="n">butterfly_sam_url</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;https://github.com/MatthewHiggins2017/BIO773P/blob/main/docs/data/butterfly_sample.csv&quot;</span>
<span class="n">butterfly_sample</span><span class="w">  </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.csv</span><span class="p">(</span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">butterfly_sam_url</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>

<span class="n">butterfly_ref_url</span><span class="w">   </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;https://github.com/MatthewHiggins2017/BIO773P/blob/main/docs/data/butterfly_reference.csv&quot;</span>
<span class="n">butterfly_reference</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.csv</span><span class="p">(</span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">butterfly_ref_url</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
<p>The <code>butterfly_reference</code> data frame contains the species name and the common name of a number of butterflies.</p>
<p>The <code>butterfly_sample</code> data frame contains information on butterflies caught in sweep netting surveys in two locations under different pesticide treatments (locations A and B). This data was collected by multiple people, who have recorded the common names of the species that they encountered (without using a standard letter case). In order to be able to compare the diversity between the two different sites, you will need to standardise the names.</p>
<h4 id="q5-write-a-function-that-adds-a-new-column-in-the-data-frame-that-contains-the-correct-latin-species-name-for-each-record-in-the-butterfly_sample">Q5 Write a function that adds a new column in the data frame that contains the correct Latin species name for each record in the <code>butterfly_sample</code>.</h4>
<p>TIP: There are several ways to do this. Remember that R is case sensitive, so you will need to account for case differences in your function. <code>grep</code> and <code>gsub</code> both allow you to set an <code>ignore.case = TRUE</code> option. Alternately, you could use the R commands <code>toupper()</code> and <code>tolower()</code>. Use the help pages to see how these work, which you can access by typing a question mark before the command - <code>?toupper</code>.</p>
<h4 id="q6-from-the-data-above-which-location-has-the-greatest-number-of-different-species">Q6. From the data above, which location has the greatest number of different species?</h4>
<h4 id="q7-which-genus-has-been-caught-the-greatest-number-of-times">Q7. Which genus has been caught the greatest number of times?</h4>
<h1 id="bonus-hacker-question">Bonus hacker question</h1>
<p>This question is an extension of the question Q5. Again, we will analyse a data frame containing the species of butterflies observed in two locations (A and B). This time, however, some of the people who did the sampling recorded the common name of the species, while others recorded the latin name (all with inconsistent letter case).</p>
<h4 id="q8-write-a-script-that-adds-a-new-column-in-the-data-frame-that-contains-the-correct-latin-species-name-for-each-record-in-the-data-frame-imported-below">Q8 Write a script that adds a new column in the data frame that contains the correct Latin species name for each record in the data frame imported below.</h4>
<div class="highlight"><pre><span></span><code><span class="n">butterfly_sam_bonus_url</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;https://github.com/MatthewHiggins2017/BIO773P/blob/main/docs/data/butterfly_sample_bonus.csv&quot;</span>
<span class="n">butterfly_sample_bonus</span><span class="w">  </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.csv</span><span class="p">(</span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">butterfly_sam_bonus_url</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
<p>One of the difficulties of this exercise is that you will have to perform a different process depending on whether the sample already has its Latin name or not. You may find that using <code>if</code> statements may be helpful (which you will have to look up). Alternatively, you may want to subset the data into the two groups (where the transformation is either from Latin to Latin or common to Latin), and do the transformation independently on each.</p>
<p>This question is not easy! But it is typical of the sort of thing researchers do from day to day and a very good test of all the things you have learned this year!</p>
<h2 id="play-time-you-can-safely-ignore-this-if-you-are-in-3rd-year">Play time (you can safely ignore this if you are in 3rd year)</h2>
<h4 id="q9-write-a-script-for-a-rock-scissors-paper-game-to-play-against-the-computer-you-will-need-to-write-it-as-a-function-eg-rock_scissors_paperyour_play">Q9 Write a script for a <a href="https://en.wikipedia.org/wiki/Rock%E2%80%93paper%E2%80%93scissors">Rock-Scissors-Paper game</a> to play against the computer. You will need to write it as a function (e.g. <code>rock_scissors_paper(your_play)</code>).</h4>
<p>NOTE: You will need to use if else statements in R. Have a look online to see how they work.</p>
<h4 id="q10-write-a-script-for-a-functional-conways-game-of-life-this-is-a-type-of-cellular-automata-model-where-very-simple-initial-rules-can-generate-highly-complex-patterns-to-modelsimulate-life-you-will-need-to-integrate-conways-rules-into-r">Q10 Write a script for a functional <a href="https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life#Rules">Conway's Game of Life</a>. This is a type of cellular automata model where very simple initial rules can generate highly complex patterns. To model/simulate life, you will need to integrate Conway's rules into R.</h4>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["content.tooltips", "navigation.instant", "navigation.tabs", "toc.integrate", "content.tabs.link"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.92b07e13.min.js"></script>
      
    
  </body>
</html>